internal/providers/azure/azure.go

```diff
@@ -71,7 +71,18 @@ func FetchFromOvfDevice(f *resource.Fetcher, ovfFsTypes []string) (types.Config,
 	logger := f.Logger
 	logger.Debug("waiting for config DVD...")
 
-	waitForCdrom(logger, devicePath)
+	// Look for the gen2 settings if avaialble
+	// else fall back to gen1 settings
+	device, err := execUtil.GetUdfBlockDevices()
+	if err != nil {
+		logger.Info("falling back to gen1 settings")
+		waitForCdrom(logger, devicePath)
+	} else if len(device) > 0 {
+		for _, dev := range device {
+			devicePath = dev
+			waitForCdrom(logger, devicePath)
```

Comment:
This doesn't make sense here.  If the device was found by `GetUdfBlockDevices()`, we've already been able to read it and determine that it has a UDF filesystem.  Instead, we should be calling `GetUdfBlockDevices()` in a retry loop until we find the correct device.  (Maybe the loop should ignore devices we've already checked, so we don't repeatedly remount an irrelevant UDF filesystem.)

Reply:
We're planning to simplify this logic by detecting the `hyperVgeneration` field for the given VM ie. which gen we're on

Reply:
Is there an out-of-band way to know that?  I am sure it's possible from the external API, and it'd likely part of the metadata passed to the VM, but we're trying to find the metadata for the VM.

It looks like there's some relevant cloud-init code over here: https://github.com/canonical/cloud-init/blob/fbcb224bc12495ba200ab107246349d802c5d8e6/cloudinit/sources/DataSourceAzure.py#L445

AFAICS it just grabs all devices of either "iso9660" or "udf".

Reply:
> Is there an out-of-band way to know that? I am sure it's possible from the external API, and it'd likely part of the metadata passed to the VM, but we're trying to find the metadata for the VM.

I'm looking at different options. So far, I came across the Azure SDK to fetch the VM details, however, that requires vendoring in a whole library.

Reply:
We've decided to use the `gen2` logic on `gen1` as well since we care about only `UDF` devices so it makes sense to go that route